File: mac-app/QuartzTeachers/Models/SectionAdder.swift
import Foundation

/// Creates a brand-new section for an existing course — the reverse of
/// removing one. The scaffolding mirrors what the setup wizard builds for
/// each section: a `sectionN/` folder with its `index.md`, plus whichever
/// per-section folders and files the course's settings call for. Shared
/// folders and files already exist at the course level and are left alone.
enum SectionAdder {

    // MARK: - Types

    /// Why a section could not be added, in words a teacher can act on.
    enum Problem: LocalizedError {
        case sectionAlreadyListed(String, Int)
        case folderInTheWay(String, Int)

        var errorDescription: String? {
            switch self {
            case .sectionAlreadyListed(let code, let number):
                return "Section \(number) of \(code) already exists."
            case .folderInTheWay(let code, let number):
                return "A folder for section \(number) of \(code) is already on disk. Move it aside first — it may hold work you want to keep."
            }
        }
    }

    // MARK: - Functions

    /// Adds one section: scaffolds its folder and adds its number to the
    /// course's settings, keeping the numbers in order. Nothing is ever
    /// written over — a folder already at the destination stops the add,
    /// because whatever is in the way may be newer work.
    static func addSection(_ sectionNumber: Int, to course: Course) throws {
        if course.sectionNumbers.contains(sectionNumber) {
            throw Problem.sectionAlreadyListed(course.code, sectionNumber)
        }

        let fileManager: FileManager = FileManager.default
        let sectionURL: URL = course.sectionDirectoryURL(forSection: sectionNumber)
        if fileManager.fileExists(atPath: sectionURL.path) {
            throw Problem.folderInTheWay(course.code, sectionNumber)
        }

        let created: String = timestamp()
        try fileManager.createDirectory(at: sectionURL, withIntermediateDirectories: true)

        // A new section should behave like the sections beside it — the same
        // pages published, the same pages held back, the same
        // display flags. So each scaffolded file copies its frontmatter from
        // the matching file in an existing section when there is one, and
        // only falls back to the wizard's plain template when there is not.
        let indexFrontmatter: String = scaffoldFrontmatter(
            sibling: siblingFile(named: "index.md", in: course),
            replacingTitleWith: sectionTitle(for: course, sectionNumber: sectionNumber),
            created: created
        )
        let indexBody: String = """
        ---
        \(indexFrontmatter)
        ---
        """
        try indexBody.write(to: sectionURL.appendingPathComponent("index.md"), atomically: true, encoding: .utf8)

        for folderName in course.configuration.perSectionFolders {
            let folderURL: URL = sectionURL.appendingPathComponent(folderName)
            try fileManager.createDirectory(at: folderURL, withIntermediateDirectories: true)
            let folderFrontmatter: String = scaffoldFrontmatter(
                sibling: siblingFile(named: "\(folderName)/index.md", in: course),
                replacingTitleWith: nil,
                fallbackTitle: folderName,
                created: created
            )
            let folderIndex: String = """
            ---
            \(folderFrontmatter)
            ---
            This is the **\(folderName)** folder. Add Markdown files to this folder to build out your site.
            """
            try folderIndex.write(to: folderURL.appendingPathComponent("index.md"), atomically: true, encoding: .utf8)
        }

        for fileName in course.configuration.perSectionFiles {
            let fileFrontmatter: String = scaffoldFrontmatter(
                sibling: siblingFile(named: fileName, in: course),
                replacingTitleWith: nil,
                fallbackTitle: fileName.replacingOccurrences(of: ".md", with: ""),
                fallbackIsDraft: SectionAdder.unpublishedFileNames.contains(fileName),
                created: created
            )
            let fileBody: String = """
            ---
            \(fileFrontmatter)
            ---
            This is the per-section file **\(fileName)**.
            """
            try fileBody.write(to: sectionURL.appendingPathComponent(fileName), atomically: true, encoding: .utf8)
        }

        // Course-level pages are shared by every section, but each section
        // decides for itself when a page appeared and whether it is
        // published — that is what `createdSectionN` / `publishForSectionN` are
        // for. A page that has those keys for the existing sections needs a
        // pair for this one too, or the new section builds it with no date
        // and no publishing state at all.
        extendCourseLevelPages(in: course, toInclude: sectionNumber, created: created)

        // Only once the folder is safely in place does the section join the
        // course's settings — the same order restore uses, so a failure
        // partway never leaves the settings pointing at nothing.
        var numbers: [Int] = course.sectionNumbers
        numbers.append(sectionNumber)
        numbers.sort()
        course.configuration.setSectionNumbers(numbers)
        try course.configuration.write(to: course.configFileURL)
    }

    /// Every markdown page at the course level — the shared folders and
    /// files, everything outside the `sectionN` folders — gains a
    /// `createdSectionN` / `publishForSectionN` pair for the section being added.
    ///
    /// The new section takes the LOWEST existing section's publishing state,
    /// for the same reason the scaffolded files copy a sibling's: a section
    /// added later should behave like the sections beside it. The date is
    /// fresh, because the page did not appear in this section until now.
    /// Pages with no per-section keys are left alone — a plain `created:`
    /// already applies to every section, including this one.
    static func extendCourseLevelPages(in course: Course, toInclude sectionNumber: Int, created: String) {
        let sectionFolderNames: Set<String> = Set(course.sectionNumbers.map { "section\($0)" })
        let fileManager: FileManager = FileManager.default
        guard let enumerator = fileManager.enumerator(
            at: course.directoryURL, includingPropertiesForKeys: nil
        ) else {
            return
        }
        while let entry = enumerator.nextObject() as? URL {
            if sectionFolderNames.contains(entry.lastPathComponent) {
                enumerator.skipDescendants()
                continue
            }
            if entry.pathExtension != "md" {
                continue
            }
            extendFrontmatter(ofPageAt: entry, toInclude: sectionNumber, created: created)
        }
    }

    /// One page's frontmatter, given a pair for the new section. Only the
    /// frontmatter block is read: a `publish: false` shown inside a fenced code
    /// block further down the page is documentation, not metadata.
    static func extendFrontmatter(ofPageAt url: URL, toInclude sectionNumber: Int, created: String) {
        guard let text = try? String(contentsOf: url, encoding: .utf8),
              let lines = frontmatterLines(ofFileAt: url) else {
            return
        }
        if alreadyHasKeys(for: sectionNumber, in: lines) {
            return
        }

        var lowestSection: Int? = nil
        for line in lines {
            guard let number = perSectionKeyNumber(in: line) else {
                continue
            }
            if lowestSection == nil || number < lowestSection! {
                lowestSection = number
            }
        }
        guard let source = lowestSection else {
            return
        }

        var addition: [String] = ["createdSection\(sectionNumber): \(created)"]
        if let publish = publishValue(forSection: source, in: lines) {
            addition.append("publishForSection\(sectionNumber): \(publish)")
        }

        // The pair goes after the last per-section key, so each section's
        // lines stay together and in order.
        var updated: [String] = []
        var lastKeyIndex: Int = -1
        for (index, line) in lines.enumerated() where perSectionKeyNumber(in: line) != nil {
            lastKeyIndex = index
        }
        for (index, line) in lines.enumerated() {
            updated.append(line)
            if index == lastKeyIndex {
                updated.append(contentsOf: addition)
            }
        }

        let body: String = String(text.dropFirst(("---\n" + lines.joined(separator: "\n")).count))
        let rewritten: String = "---\n" + updated.joined(separator: "\n") + body
        try? rewritten.write(to: url, atomically: true, encoding: .utf8)
    }

    /// Does this page already carry the new section's keys?
    static func alreadyHasKeys(for sectionNumber: Int, in lines: [String]) -> Bool {
        let prefixes: [String] = [
            "createdSection\(sectionNumber):",
            "publishForSection\(sectionNumber):",
            "draftSection\(sectionNumber):",
        ]
        for line in lines {
            for prefix in prefixes {
                if line.hasPrefix(prefix) {
                    return true
                }
            }
        }
        return false
    }

    /// Whether a section publishes this page, as the string to write back.
    ///
    /// Visibility is `publishForSectionN`. `draftSectionN` is the older
    /// spelling with the OPPOSITE polarity, so a course written before the
    /// rename is read and inverted — carrying it across unchanged would
    /// publish a page the teacher had held back.
    static func publishValue(forSection sectionNumber: Int, in lines: [String]) -> String? {
        let publishPrefix: String = "publishForSection\(sectionNumber):"
        for line in lines where line.hasPrefix(publishPrefix) {
            return String(line.dropFirst(publishPrefix.count)).trimmingCharacters(in: .whitespaces)
        }

        let draftPrefix: String = "draftSection\(sectionNumber):"
        for line in lines where line.hasPrefix(draftPrefix) {
            let value: String = String(line.dropFirst(draftPrefix.count)).trimmingCharacters(in: .whitespaces)
            return value.lowercased() == "true" ? "false" : "true"
        }
        return nil
    }

    /// The section number in a per-section key.
    static func perSectionKeyNumber(in line: String) -> Int? {
        for prefix in ["createdSection", "publishForSection", "draftSection"] {
            guard line.hasPrefix(prefix) else {
                continue
            }
            let rest: Substring = line.dropFirst(prefix.count)
            guard let colon = rest.firstIndex(of: ":") else {
                continue
            }
            return Int(rest[rest.startIndex..<colon])
        }
        return nil
    }

    /// The same file in an existing section, if any section has it — the
    /// lowest-numbered one wins, so the choice is predictable.
    static func siblingFile(named relativePath: String, in course: Course) -> URL? {
        var numbers: [Int] = course.sectionNumbers
        numbers.sort()
        for number in numbers {
            let candidate: URL = course.sectionDirectoryURL(forSection: number).appendingPathComponent(relativePath)
            if FileManager.default.fileExists(atPath: candidate.path) {
                return candidate
            }
        }
        return nil
    }

    /// Pages that exist for the teacher's own eyes — the wizard creates
    /// them held back so they are never published, and a section added
    /// with no sibling to imitate must do the same.
    static let unpublishedFileNames: Set<String> = ["Private Notes.md", "Scratch Page.md"]

    /// The frontmatter for one scaffolded file. A sibling's frontmatter is
    /// carried over whole — publish status, table-of-contents and backlink
    /// flags, anything else the teacher set — with only `created:` freshened
    /// and, when a new title is given, the title swapped. Without a sibling,
    /// the wizard's plain template stands in.
    static func scaffoldFrontmatter(
        sibling: URL?,
        replacingTitleWith newTitle: String?,
        fallbackTitle: String? = nil,
        fallbackIsDraft: Bool = false,
        created: String
    ) -> String {
        if let sibling, let siblingLines = frontmatterLines(ofFileAt: sibling) {
            var result: [String] = []
            for line in siblingLines {
                if line.hasPrefix("created:") {
                    result.append("created: \(created)")
                } else if line.hasPrefix("title:"), let newTitle {
                    result.append("title: \(newTitle)")
                } else {
                    result.append(line)
                }
            }
            return result.joined(separator: "\n")
        }
        let title: String = newTitle ?? fallbackTitle ?? ""
        let publish: String = fallbackIsDraft ? "false" : "true"
        return "title: \(title)\ncreated: \(created)\npublish: \(publish)"
    }

    /// The lines between a file's opening and closing `---` markers, or nil
    /// when the file has no frontmatter to speak of.
    static func frontmatterLines(ofFileAt url: URL) -> [String]? {
        guard let text = try? String(contentsOf: url, encoding: .utf8) else {
            return nil
        }
        let lines: [String] = text.components(separatedBy: "\n")
        guard lines.first == "---" else {
            return nil
        }
        var collected: [String] = []
        var index: Int = 1
        while index < lines.count {
            if lines[index] == "---" {
                return collected
            }
            collected.append(lines[index])
            index += 1
        }
        return nil
    }

    /// The new section's landing-page title. A sibling section's title with
    /// the section number swapped is the most faithful source — it keeps
    /// whatever wording the teacher chose. Failing that, the wizard's form,
    /// without doubling the grade when the course name already leads with it
    /// ("Grade 9 Science" must not become "Grade 9 Grade 9 Science").
    static func sectionTitle(for course: Course, sectionNumber: Int) -> String {
        if let siblingIndex = siblingFile(named: "index.md", in: course),
           let siblingLines = frontmatterLines(ofFileAt: siblingIndex) {
            for line in siblingLines {
                if line.hasPrefix("title:") {
                    let siblingTitle: String = String(line.dropFirst("title:".count)).trimmingCharacters(in: .whitespaces)
                    if let range = siblingTitle.range(of: #"Section \d+$"#, options: .regularExpression) {
                        return siblingTitle.replacingCharacters(in: range, with: "Section \(sectionNumber)")
                    }
                }
            }
        }
        let gradeLabel: String = SectionAdder.gradeLabel(forCourseCode: course.code)
        let courseName: String = course.configuration.courseName
        // Deliberately literal, matching the build: the switch alone
        // decides. The settings warn when the name already carries the
        // grade; resolving that is the teacher's call, never guessed.
        var titlePrefix: String = ""
        if course.configuration.showsGradeInTitle(forSection: sectionNumber) && !gradeLabel.isEmpty {
            titlePrefix = "\(gradeLabel) "
        }
        return "\(titlePrefix)\(courseName), Section \(sectionNumber)"
    }

    /// The grade named by the course code's fourth character, matching the
    /// wizard: "ICS3U" → "Grade 11". A club code like "CODING" has no grade
    /// digit there, so no grade label at all.
    static func gradeLabel(forCourseCode code: String) -> String {
        let characters: [Character] = Array(code)
        guard characters.count >= 4 else {
            return ""
        }
        let gradeCharacter: Character = characters[3]
        guard gradeCharacter.isNumber else {
            return ""
        }
        switch gradeCharacter {
        case "1": return "Grade 9"
        case "2": return "Grade 10"
        case "3": return "Grade 11"
        case "4": return "Grade 12"
        default: return "Grade ?"
        }
    }

    /// The `created:` timestamp, in the same form the wizard writes:
    /// `2026-08-10T14:30:00.000-0400`.
    static func timestamp(for date: Date = Date()) -> String {
        let formatter: DateFormatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd'T'HH:mm:ss.000Z"
        return formatter.string(from: date)
    }

    /// The number the sheet should offer first: the smallest section number
    /// not already in use, so "add section 1" and "add the next one" are
    /// each a single click.
    static func suggestedNumber(existing: [Int]) -> Int {
        var candidate: Int = 1
        while existing.contains(candidate) {
            candidate += 1
        }
        return candidate
    }

    /// What is wrong with the typed section number, or nil when nothing is —
    /// worded like the wizard's own warnings. An empty field is quietly
    /// invalid: nothing has been said yet, so there is nothing to warn about.
    static func entryProblem(_ entry: String, existing: [Int], courseCode: String) -> String? {
        let trimmed: String = entry.trimmingCharacters(in: .whitespaces)
        if trimmed.isEmpty {
            return nil
        }
        guard let number = Int(trimmed) else {
            return "“\(trimmed)” isn’t a section number — sections are 1 or higher."
        }
        if number < 1 {
            return "“\(trimmed)” isn’t a section number — sections are 1 or higher."
        }
        if existing.contains(number) {
            return "Section \(number) of \(courseCode) already exists."
        }
        return nil
    }

    /// True when the typed entry names a section that can be added right now.
    static func entryIsAddable(_ entry: String, existing: [Int]) -> Bool {
        let trimmed: String = entry.trimmingCharacters(in: .whitespaces)
        guard let number = Int(trimmed) else {
            return false
        }
        return number >= 1 && !existing.contains(number)
    }
}

File: windows-app/Plantoir.Core/Models/SectionAdder.cs
using System.Globalization;

namespace Plantoir.Core.Models;

/// <summary>
/// Scaffolds a brand-new section for an existing course, imitating a sibling
/// section where one exists so a page the siblings keep unpublished starts
/// unpublished here too.
/// </summary>
public static class SectionAdder
{
    /// <summary>Teacher-eyes-only pages: created unpublished, never in the site.</summary>
    public static readonly IReadOnlySet<string> UnpublishedFileNames =
        new HashSet<string> { "Private Notes.md", "Scratch Page.md" };

    public sealed class SectionAddException(string message) : Exception(message);

    public static void AddSection(int number, Course course)
    {
        var config = course.Configuration;
        if (config.SectionNumbers.Contains(number))
            throw new SectionAddException($"Section {number} of {course.Code} already exists.");
        string sectionDir = course.SectionDirectory(number);
        if (Directory.Exists(sectionDir))
            throw new SectionAddException(
                $"A folder for section {number} of {course.Code} is already on disk. Move it aside first — it may hold work you want to keep.");

        string created = Timestamp(DateTimeOffset.Now);
        Directory.CreateDirectory(sectionDir);

        // index.md — frontmatter only, title recomputed for this section.
        string indexFrontmatter = ScaffoldFrontmatter(
            SiblingFile("index.md", course),
            newTitle: SectionTitle(course, number),
            fallbackTitle: null, fallbackIsDraft: false, created: created);
        File.WriteAllText(Path.Combine(sectionDir, "index.md"), $"---\n{indexFrontmatter}\n---");

        foreach (string folder in config.PerSectionFolders)
        {
            string folderPath = Path.Combine(sectionDir, folder);
            Directory.CreateDirectory(folderPath);
            string fm = ScaffoldFrontmatter(
                SiblingFile(folder + "/index.md", course),
                newTitle: null, fallbackTitle: folder, fallbackIsDraft: false, created: created);
            File.WriteAllText(Path.Combine(folderPath, "index.md"),
                $"---\n{fm}\n---\nThis is the **{folder}** folder. Add Markdown files to this folder to build out your site.");
        }

        foreach (string fileName in config.PerSectionFiles)
        {
            string fm = ScaffoldFrontmatter(
                SiblingFile(fileName, course),
                newTitle: null,
                fallbackTitle: fileName.Replace(".md", ""),
                fallbackIsDraft: UnpublishedFileNames.Contains(fileName),
                created: created);
            File.WriteAllText(Path.Combine(sectionDir, fileName),
                $"---\n{fm}\n---\nThis is the per-section file **{fileName}**.");
        }

        // Only after the folder is safely written does the config learn about
        // it — a mid-way failure never leaves settings pointing at nothing.
        var numbers = config.SectionNumbers;
        numbers.Add(number);
        numbers.Sort();
        config.SetSectionNumbers(numbers);
        config.Write(course.ConfigFilePath);
    }

    /// <summary>The first existing copy among ascending section numbers — deterministic.</summary>
    internal static string? SiblingFile(string relativePath, Course course)
    {
        foreach (int n in course.Configuration.SectionNumbers.OrderBy(n => n))
        {
            string candidate = Path.Combine(course.SectionDirectory(n),
                relativePath.Replace('/', Path.DirectorySeparatorChar));
            if (File.Exists(candidate)) return candidate;
        }
        return null;
    }

    /// <summary>
    /// Lines between an opening "---" (which must be line 1, exactly) and the
    /// closing "---". No closing marker → null.
    /// </summary>
    internal static List<string>? FrontmatterLines(string path)
    {
        string text;
        try { text = File.ReadAllText(path); } catch { return null; }
        var lines = text.Split('\n');
        if (lines.Length == 0 || lines[0].TrimEnd('\r') != "---") return null;
        var collected = new List<string>();
        for (int i = 1; i < lines.Length; i++)
        {
            if (lines[i].TrimEnd('\r') == "---") return collected;
            collected.Add(lines[i].TrimEnd('\r'));
        }
        return null;
    }

    /// <summary>
    /// A sibling's frontmatter carried over whole, with only the created date
    /// freshened and (for index.md) the title replaced.
    /// </summary>
    internal static string ScaffoldFrontmatter(string? siblingPath, string? newTitle,
                                               string? fallbackTitle, bool fallbackIsDraft, string created)
    {
        var siblingLines = siblingPath is null ? null : FrontmatterLines(siblingPath);
        if (siblingLines is not null)
        {
            var rewritten = siblingLines.Select(line =>
            {
                if (line.StartsWith("created:", StringComparison.Ordinal)) return "created: " + created;
                if (line.StartsWith("title:", StringComparison.Ordinal) && newTitle is not null) return "title: " + newTitle;
                return line;
            });
            return string.Join("\n", rewritten);
        }
        // The publish key, not the legacy draft one, and note the polarity is
        // INVERTED: a teacher-eyes-only page is `publish: false`. Missing this
        // would have quietly born every new section in the old schema — the
        // same gap that setup_course.py had, in the other creation path.
        //
        // Only the fallback writes this. When there is a sibling section its
        // frontmatter is copied verbatim, which is what a course still using
        // `draft:` wants: the new section matches its siblings rather than
        // becoming the one page in the course with a different vocabulary.
        string title = newTitle ?? fallbackTitle ?? "";
        return $"title: {title}\ncreated: {created}\npublish: {(fallbackIsDraft ? "false" : "true")}";
    }

    /// <summary>
    /// Prefer the sibling's own title with its trailing "Section N" renumbered
    /// (preserving the teacher's wording); otherwise the wizard's form. The
    /// grade prefix is LITERAL — the switch alone decides.
    /// </summary>
    public static string SectionTitle(Course course, int sectionNumber)
    {
        string? sibling = SiblingFile("index.md", course);
        if (sibling is not null && FrontmatterLines(sibling) is { } lines)
        {
            foreach (string line in lines)
            {
                if (!line.StartsWith("title:", StringComparison.Ordinal)) continue;
                string value = line["title:".Length..].Trim();
                var match = System.Text.RegularExpressions.Regex.Match(value, @"Section \d+$");
                if (match.Success)
                    return value[..match.Index] + "Section " + sectionNumber;
                break;
            }
        }
        string gradeLabel = GradeLabel(course.Code);
        bool showsGrade = course.Configuration.ShowsGradeInTitle(sectionNumber);
        string prefix = showsGrade && gradeLabel.Length > 0 ? gradeLabel + " " : "";
        return $"{prefix}{course.Configuration.CourseName}, Section {sectionNumber}";
    }

    /// <summary>The 4th character of an Ontario course code carries its grade.</summary>
    public static string GradeLabel(string courseCode)
    {
        if (courseCode.Length < 4) return "";
        char c = courseCode[3];
        if (!char.IsDigit(c)) return "";
        return c switch
        {
            '1' => "Grade 9",
            '2' => "Grade 10",
            '3' => "Grade 11",
            '4' => "Grade 12",
            _ => "Grade ?",
        };
    }

    /// <summary>Wizard-style created stamp: 2026-08-10T14:30:00.000-0400 (offset without colon).</summary>
    public static string Timestamp(DateTimeOffset date)
    {
        string body = date.ToString("yyyy-MM-dd'T'HH:mm:ss.'000'", CultureInfo.InvariantCulture);
        string offset = date.ToString("zzz", CultureInfo.InvariantCulture).Replace(":", "");
        return body + offset;
    }

    /// <summary>The smallest positive number not already in use.</summary>
    public static int SuggestedNumber(IReadOnlyCollection<int> existing)
    {
        int candidate = 1;
        while (existing.Contains(candidate)) candidate++;
        return candidate;
    }

    /// <summary>Live validation for the Add Section sheet. Empty entry → no warning yet.</summary>
    public static string? EntryProblem(string entry, IReadOnlyCollection<int> existing, string courseCode)
    {
        string trimmed = entry.Trim();
        if (trimmed.Length == 0) return null;
        if (!int.TryParse(trimmed, out int number) || number < 1)
            return $"“{trimmed}” isn’t a section number — sections are 1 or higher.";
        if (existing.Contains(number))
            return $"Section {number} of {courseCode} already exists.";
        return null;
    }

    public static bool EntryIsAddable(string entry, IReadOnlyCollection<int> existing) =>
        int.TryParse(entry.Trim(), out int number) && number >= 1 && !existing.Contains(number);
}
